crates/agglayer-node/src/rate_limiting/local/mod.rs

```diff
@@ -0,0 +1,63 @@
+//! Local rate limiter, i.e. one constraining a single network.
+
+use agglayer_config::rate_limiting::TimeRateLimit;
+use tokio::time::Instant;
+
+// TODO: make private if possible
+mod interface;
+pub mod trivial;
+pub mod wall_clock;
+
+pub use interface::RateLimiter;
+
+pub struct LocalRateLimiter {
+    /// Rate limiter for `sendTx` settlement.
+    send_tx: trivial::RateLimiter<wall_clock::RateLimiter>,
```

Comment:
I'm wondering if we could use a custom struct such as `SendTx {}` and `SendCertificate`, and have a dynamic hashmap where we use a `TypeId` on that struct. So that we can call `rate_limiter.check::<SendTx>(Instant::now())`. Maybe it's a bad idea, or not even possible, I didn't tested it.

Reply:
> I'm wondering if we could use a custom struct such as `SendTx {}` and `SendCertificate`, and have a dynamic hashmap where we use a `TypeId` on that struct. So that we can call `rate_limiter.check::<SendTx>(Instant::now())`. Maybe it's a bad idea, or not even possible, I didn't tested it.

I think it is possible but in my opinion not worth all the trait magic needed to accomplish this. Also the implementation for the epoch-based rate limiting will have a different (more efficient) internal representation since it does not need to track the list of events in given window, just the number of events in the current epoch.

Reply:
Yeah, I was wondering how the api of the LocalRateLimiter would look like if we had 10 or 20 RPC, but we have time to think about that